File: javascript/timehash.js
/*
 * Timehash
 * Copyright (c) 2026-present NAVER Corp.
 * Apache-2.0
 */

/**
 * Timehash: Hierarchical time indexing for efficient business hours search.
 * 
 * Timehash generates hierarchical hash keys for time ranges, enabling efficient
 * temporal filtering in search systems with minimal index overhead.
 */

class Timehash {
    /**
     * Default measures: [400, 100, 15, 5, 1] representing:
     * - 400 minutes (4 hours)
     * - 100 minutes (1 hour, treated as 60 minutes)
     * - 15 minutes
     * - 5 minutes
     * - 1 minute
     */
    static DEFAULT_MEASURES = [400, 100, 15, 5, 1];

    /**
     * Initialize Timehash with optional custom measures.
     * 
     * @param {number[]} measures - List of time measures in minutes. Default: [400, 100, 15, 5, 1]
     */
    constructor(measures = null) {
        this.measures = measures || [...Timehash.DEFAULT_MEASURES];
    }

    /**
     * Generate query terms for a given time.
     * 
     * @param {number} hhmm - Time in hhmm format (e.g., 1720 for 17:20)
     * @returns {string[]} List of hierarchical hash keys for the query time
     */
    getQueryTerms(hhmm) {
        const terms = [];

        if (hhmm > 10000) {
            return terms;
        }

        for (const measure of this.measures) {
            let target = hhmm;
            let measureVal;

            if (measure >= 100) {
                target = Math.floor(target / 100);
                measureVal = Math.floor(measure / 100);
            } else {
                target = target % 100;
                measureVal = measure;
            }

            const key = Math.floor(target / measureVal) * measureVal;
            const term = this._twoDigitTerm(key);

            if (terms.length === 0) {
                terms.push(term);
            } else {
                terms.push(terms[terms.length - 1] + term);
            }
        }

        return terms;
    }

    /**
     * Generate index terms for a time range.
     * 
     * @param {string} fromTime - Start time in hhmm format (e.g., "1140")
     * @param {string} toTime - End time in hhmm format (e.g., "2100")
     * @returns {string[]} List of hierarchical hash keys covering the time range
     */
    getIndexTerms(fromTime, toTime) {
        const terms = [];
        const fromInt = parseInt(fromTime, 10);
        const toInt = parseInt(toTime, 10);
        this._cover(terms, 0, "", fromInt, toInt);
        return terms;
    }

    /**
     * Convert a key to two-digit string representation.
     * 
     * @param {number} key - Key value
     * @returns {string} Two-digit string representation
     * @private
     */
    _twoDigitTerm(key) {
        let k = key;
        if (k >= 100) {
            k = Math.floor(k / 100);
        }
        return `${Math.floor(k / 10)}${k % 10}`;
    }

    /**
     * Recursively generate keys for a time range.
     * 
     * @param {string[]} terms - List to append generated keys
     * @param {number} measureIndex - Current level in the hierarchy
     * @param {string} parentKey - Key from parent level
     * @param {number} fromTime - Start time (integer in hhmm format)
     * @param {number} toTime - End time (integer in hhmm format)
     * @private
     */
    _cover(terms, measureIndex, parentKey, fromTime, toTime) {
        if (measureIndex >= this.measures.length) {
            return;
        }

        const measure = this.measures[measureIndex];
        let fromVal = fromTime;
        let toVal = toTime;

        if (measure < 100) {
            fromVal = fromVal % 100;
            toVal = toVal % 100;
            if (toVal === 0) {
                toVal = 60;
            }
        }

        for (let i = Math.floor(fromVal / measure); i <= Math.floor(toVal / measure); i++) {
            const key = measure * i;

            if (fromVal <= key && key + measure <= toVal) {
                terms.push(parentKey + this._twoDigitTerm(key));
            } else if (key < fromVal && fromVal < key + measure) {
                const end = Math.min(key + measure, toVal);
                this._cover(terms, measureIndex + 1,
                    parentKey + this._twoDigitTerm(key),
                    fromVal, end);
            } else if (key < toVal && toVal < key + measure) {
                this._cover(terms, measureIndex + 1,
                    parentKey + this._twoDigitTerm(key),
                    key, toVal);
            }
        }
    }
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = Timehash;
}

// Export for ES6 modules
if (typeof window !== 'undefined') {
    window.Timehash = Timehash;
}

File: python/timehash.py
# Timehash
# Copyright (c) 2026-present NAVER Corp.
# Apache-2.0

"""
Timehash: Hierarchical time indexing for efficient business hours search.

Timehash generates hierarchical hash keys for time ranges, enabling efficient
temporal filtering in search systems with minimal index overhead.
"""

from typing import List


class Timehash:
    """
    Timehash generates hierarchical hash keys for time ranges.
    
    The default measures are [400, 100, 15, 5, 1] representing:
    - 400 minutes (4 hours)
    - 100 minutes (1 hour 40 minutes, but treated as 1 hour)
    - 15 minutes
    - 5 minutes
    - 1 minute
    """
    
    def __init__(self, measures: List[int] = None):
        """
        Initialize Timehash with optional custom measures.
        
        Args:
            measures: List of time measures in minutes. Default: [400, 100, 15, 5, 1]
        """
        self.measures = measures if measures is not None else [400, 100, 15, 5, 1]
    
    def get_query_terms(self, hhmm: int) -> List[str]:
        """
        Generate query terms for a given time.
        
        Args:
            hhmm: Time in hhmm format (e.g., 1720 for 17:20)
            
        Returns:
            List of hierarchical hash keys for the query time
        """
        terms = []
        
        if hhmm > 10000:
            return terms
        
        for measure in self.measures:
            target = hhmm
            
            if measure >= 100:
                target //= 100
                measure_val = measure // 100
            else:
                target %= 100
                measure_val = measure
            
            key = (target // measure_val) * measure_val
            term = self._two_digit_term(key)
            
            if not terms:
                terms.append(term)
            else:
                terms.append(terms[-1] + term)
        
        return terms
    
    def get_index_terms(self, from_time: str, to_time: str) -> List[str]:
        """
        Generate index terms for a time range.
        
        Args:
            from_time: Start time in hhmm format (e.g., "1140")
            to_time: End time in hhmm format (e.g., "2100")
            
        Returns:
            List of hierarchical hash keys covering the time range
        """
        terms = []
        from_int = int(from_time)
        to_int = int(to_time)
        self._cover(terms, 0, "", from_int, to_int)
        return terms
    
    def _two_digit_term(self, key: int) -> str:
        """Convert a key to two-digit string representation."""
        if key >= 100:
            key //= 100
        
        return f"{key // 10}{key % 10}"
    
    def _cover(self, terms: List[str], measure_index: int, parent_key: str, 
               from_time: int, to_time: int) -> None:
        """
        Recursively generate keys for a time range.
        
        Args:
            terms: List to append generated keys
            measure_index: Current level in the hierarchy
            parent_key: Key from parent level
            from_time: Start time (integer in hhmm format)
            to_time: End time (integer in hhmm format)
        """
        if measure_index >= len(self.measures):
            return
        
        measure = self.measures[measure_index]
        from_val = from_time
        to_val = to_time
        
        if measure < 100:
            from_val %= 100
            to_val %= 100
            if to_val == 0:
                to_val = 60
        
        for i in range(from_val // measure, (to_val // measure) + 1):
            key = measure * i
            
            if from_val <= key and key + measure <= to_val:
                terms.append(parent_key + self._two_digit_term(key))
            elif key < from_val < key + measure:
                self._cover(terms, measure_index + 1, 
                           parent_key + self._two_digit_term(key),
                           from_val, min(key + measure, to_val))
            elif key < to_val < key + measure:
                self._cover(terms, measure_index + 1,
                           parent_key + self._two_digit_term(key),
                           key, to_val)
